File: packages/ios/Byoky/Services/UsageParser.swift
import Foundation

enum UsageParser {
    struct TokenUsage {
        let inputTokens: Int
        let outputTokens: Int
    }

    static func parseModel(from body: Data?) -> String? {
        guard let body, body.count <= 10_485_760,
              let json = try? JSONSerialization.jsonObject(with: body) as? [String: Any],
              let model = json["model"] as? String else {
            return nil
        }
        return model
    }

    static func parseUsage(providerId: String, body: String) -> TokenUsage? {
        // Streaming (SSE) — scan data: lines from the end
        if body.contains("data: ") {
            let lines = body.components(separatedBy: "\n")
                .filter { $0.hasPrefix("data: ") && !$0.contains("[DONE]") }
            for line in lines.reversed() {
                let json = String(line.dropFirst(6))
                guard let data = json.data(using: .utf8),
                      let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let usage = extractUsage(providerId: providerId, from: parsed) else {
                    continue
                }
                return usage
            }
            return nil
        }

        // Non-streaming JSON
        guard let data = body.data(using: .utf8),
              let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return extractUsage(providerId: providerId, from: parsed)
    }

    private static func extractUsage(providerId: String, from parsed: [String: Any]) -> TokenUsage? {
        // Anthropic: { usage: { input_tokens, output_tokens } }
        if providerId == "anthropic",
           let usage = parsed["usage"] as? [String: Any],
           let input = usage["input_tokens"] as? Int,
           let output = usage["output_tokens"] as? Int {
            return sanitize(input: input, output: output)
        }

        // Gemini: { usageMetadata: { promptTokenCount, candidatesTokenCount } }
        if providerId == "gemini",
           let meta = parsed["usageMetadata"] as? [String: Any],
           let prompt = meta["promptTokenCount"] as? Int {
            let candidates = meta["candidatesTokenCount"] as? Int ?? 0
            return sanitize(input: prompt, output: candidates)
        }

        // OpenAI-compatible: { usage: { prompt_tokens, completion_tokens } }
        if let usage = parsed["usage"] as? [String: Any],
           let prompt = usage["prompt_tokens"] as? Int,
           let completion = usage["completion_tokens"] as? Int {
            return sanitize(input: prompt, output: completion)
        }

        return nil
    }

    private static func sanitize(input: Int, output: Int) -> TokenUsage? {
        let i = max(0, input)
        let o = max(0, output)
        guard i >= 0, o >= 0 else { return nil }
        return TokenUsage(inputTokens: i, outputTokens: o)
    }
}

File: packages/android/app/src/main/java/com/byoky/app/data/UsageParser.kt
package com.byoky.app.data

import org.json.JSONObject

object UsageParser {

    data class TokenUsage(val inputTokens: Int, val outputTokens: Int)

    fun parseModel(body: ByteArray?): String? {
        if (body == null || body.size > 10_485_760) return null
        return try {
            val json = JSONObject(String(body, Charsets.UTF_8))
            json.optString("model", "").ifEmpty { null }
        } catch (_: Exception) {
            null
        }
    }

    fun parseUsage(providerId: String, body: String): TokenUsage? {
        // Streaming (SSE) — scan data: lines from the end
        if (body.contains("data: ")) {
            val lines = body.split("\n")
                .filter { it.startsWith("data: ") && !it.contains("[DONE]") }
            for (line in lines.reversed()) {
                val json = line.removePrefix("data: ")
                try {
                    val parsed = JSONObject(json)
                    val usage = extractUsage(providerId, parsed)
                    if (usage != null) return usage
                } catch (_: Exception) {
                    continue
                }
            }
            return null
        }

        // Non-streaming JSON
        return try {
            val parsed = JSONObject(body)
            extractUsage(providerId, parsed)
        } catch (_: Exception) {
            null
        }
    }

    private fun extractUsage(providerId: String, parsed: JSONObject): TokenUsage? {
        // Anthropic: { usage: { input_tokens, output_tokens } }
        if (providerId == "anthropic") {
            val usage = parsed.optJSONObject("usage") ?: return null
            val input = usage.optInt("input_tokens", -1)
            val output = usage.optInt("output_tokens", -1)
            if (input >= 0 && output >= 0) return sanitize(input, output)
        }

        // Gemini: { usageMetadata: { promptTokenCount, candidatesTokenCount } }
        if (providerId == "gemini") {
            val meta = parsed.optJSONObject("usageMetadata") ?: return null
            val prompt = meta.optInt("promptTokenCount", -1)
            if (prompt < 0) return null
            val candidates = meta.optInt("candidatesTokenCount", 0)
            return sanitize(prompt, candidates)
        }

        // OpenAI-compatible: { usage: { prompt_tokens, completion_tokens } }
        val usage = parsed.optJSONObject("usage") ?: return null
        val prompt = usage.optInt("prompt_tokens", -1)
        val completion = usage.optInt("completion_tokens", -1)
        if (prompt >= 0 && completion >= 0) return sanitize(prompt, completion)

        return null
    }

    private fun sanitize(input: Int, output: Int): TokenUsage? {
        val i = maxOf(0, input)
        val o = maxOf(0, output)
        return TokenUsage(i, o)
    }
}
